sim/cboria/src/bin/cbor-dump.rs

```diff
@@ -0,0 +1,115 @@
+//! Dump a CBor encoded file in a readable fashion.
+
+use byteorder::{
+    BigEndian,
+    ReadBytesExt,
+};
+use cboria::pdump::HexDump;
+use std::{
+    env,
+    fs::File,
+    io::{BufReader, Read},
+    result,
+};
+
+type Result<T> = result::Result<T, failure::Error>;
+
+fn main() -> Result<()> {
+    for arg in env::args().skip(1) {
+        println!("decoding: {:?}", arg);
+        let mut buf = BufReader::new(File::open(arg)?);
+        walk(&mut buf, Indent::new())?;
+    }
+    Ok(())
+}
+
+/// Walk a CBOR structure from the given reader.
+fn walk<R: Read>(rd: &mut R, indent: Indent) -> Result<()> {
+    let (code, count) = get_code(rd)?;
+    // println!("code: {}, count: 0x{:x}", code, count);
+    match code {
+        2 => {
+            println!("{}{:02x} {:<10x}{} Byte string", indent.left(), code, count,
+                indent.pad(2 + 1 + 10));
+            let mut buf = vec![0u8; count];
+            rd.read_exact(&mut buf)?;
+            buf.dump();
+            // Sometimes byte strings are nested CBOR elements, but this
+            // can't be determined without knowing the schema.
+        }
+        4 => {
+            println!("{}{:02x} {:<10x}{} Array", indent.left(), code, count,
+                indent.pad(2 + 1 + 10));
+            for _ in 0 .. count {
+                walk(rd, indent.succ())?;
+            }
+        }
+        5 => {
+            println!("{}{:02x} {:<10x}{} Map", indent.left(), code, count,
+                indent.pad(2 + 1 + 10));
+            for _ in 0 .. count {
+                walk(rd, indent.succ())?;
+                walk(rd, indent.succ())?;
+            }
+        }
+        6 => {
+            println!("{}{:02x} {:<10x}{} Tag", indent.left(), code, count,
+                indent.pad(2 + 1 + 10));
+            walk(rd, indent.succ())?;
+        }
+        code => panic!("TODO: Implement code {}", code),
+    }
+
+    Ok(())
+}
+
+/// Read a CBOR code value, and the associated integer.
+fn get_code<R: Read>(rd: &mut R) -> Result<(u8, usize)> {
+    let b = rd.read_u8()?;
+    let code = b >> 5;
+    let count = match b & 0x1f {
+        count @ 0 ..= 23 => count as usize,
```

Comment:
The binding should be `count @ 0 ... 23`, no? What does the `..=` syntax mean?

Reply:
`..=` is an inclusive range pattern (as opposed to `0 .. 24` which would be the same range). This is new to Rust 2018.